mentor/messages.py

```diff
@@ -0,0 +1,370 @@
+
+    @property
+    def exception(self):
+        try:
+            return remote_exception(*self.data)
+        except:
+            return None
+
+
+# TODO create custom messages per executor
+class PythonTask(PickleMixin, TaskInfo):
+
+     
+    def __init__(self, fn=None, args=[], kwargs={},
+                 resources=[Cpus(0.1), Mem(128), Disk(0)],
+                 executor=None, retries=3, **kwds):
+        super(PythonTask, self).__init__(**kwds)
+        self.status = PythonTaskStatus(task_id=self.id, state='TASK_STAGING')
+        self.executor = executor or PythonExecutor()
+        self.data = (fn, args, kwargs)
+        self.resources = resources
+        self.retries = retries
+        self.attempt = 1
+
+        self.labels = Message(labels=Message(key='python'))
+
+    def __call__(self):
+        fn, args, kwargs = self.data
+        return fn(*args, **kwargs)
+
+    def retry(self, status):
+        if self.attempt < self.retries:
+            log.info('Task {} attempt #{} rescheduled due to failure with state '
+                         '{} and message {}'.format(self.id, self.attempt,
+                                                    status.state, status.message))
+            self.attempt += 1
+            status.state = 'TASK_STAGING'
+        else:
+            log.error('Aborting due to task {} failed for {} attempts in state '
+                          '{} with message {}'.format(self.id, self.retries,
+                                                      status.state, status.message))
+            raise RuntimeError('Task {} failed with state {} and message {}'.format(
+                self.id, status.state, status.message))
+
+    def update(self, status):
+        self.on_update(status)
+        if status.has_succeeded():
+            self.on_success(status)
+        elif status.has_failed():
+            self.on_fail(status)
+
+    def on_update(self, status):
+        self.status = status  # update task's status
+        log.info('Task {} has been updated with state {}'.format(
+            self.id.value, status.state))
+
+    def on_success(self, status):
+        log.info('Task {} has been succeded'.format(self.id.value))
+
+    def on_fail(self, status):
+        log.error('Task {} has been failed with state {} due to {}'.format(
+            self.id.value, status.state, status.message))
+
+        try:
+            raise status.exception  # won't retry due to code error in PythonTaskStatus
+        except KeyError as e:
+            # not a code error, e.g. problem during deployment
+            self.retry(status)
+        else:
+            log.error('Aborting due to task {} failed with state {} and message '
+                          '{}'.format(self.id, status.state, status.message))
+
+
+class PythonExecutor(Message):
+
+ 
+
+    def __init__(self, docker='satyr', force_pull=False,
+                 envs={}, uris=[], **kwds):
+        super(PythonExecutor, self).__init__(**kwds)
```

Comment:
Maybe we should use something like a template with pure dict notation instead of Message constructors.